keras_nlp/utils/pipeline_model.py

```diff
@@ -0,0 +1,292 @@
+    )
+    validation_data = tf.nest.map_structure(
+        lambda t: None if t is None else t[split_at:], data
+    )
+    return training_data, validation_data
+
+
+class PipelineModel(keras.Model):
+    """A model which allows automatically applying preprocessing."""
+
+    def __init__(self, *args, include_preprocessing=True, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.include_preprocessing = include_preprocessing
+        # Stop SavedModel from trying to trace pipeline models, and rather
+        # always restore from config to an actual python object.
+        self._must_restore_from_config = True
+
+    def preprocess_features(self, x):
+        """An overridable function which preprocesses features."""
+        return x
+
+    def preprocess_labels(self, y):
+        """An overridable function which preprocesses labels."""
+        return y
+
+    def preprocess_samples(self, x, y=None, sample_weight=None):
+        """An overridable function which preprocesses entire samples."""
+        x = self.preprocess_features(x)
+        if y is not None:
+            y = self.preprocess_labels(y)
+        return pack_x_y_sample_weight(x, y, sample_weight)
+
+    # ========================================================================
+    # Below are overrides to keras.Model methods to apply the functions above.
+    # ========================================================================
+    def fit(
+        self,
+        x=None,
+        y=None,
+        batch_size=None,
+        sample_weight=None,
+        validation_data=None,
+        validation_split=None,
+        **kwargs,
+    ):
+        if validation_split and validation_data is None:
+            (x, y, sample_weight), validation_data = train_validation_split(
+                (x, y, sample_weight), validation_split=validation_split
+            )
+
+        x = convert_inputs_to_dataset(x, y, sample_weight, batch_size)
+        if self.include_preprocessing:
+            x = x.map(
+                self.preprocess_samples, num_parallel_calls=tf.data.AUTOTUNE
+            ).prefetch(tf.data.AUTOTUNE)
+
+        if validation_data is not None:
+            if not isinstance(validation_data, tf.data.Dataset):
+                (vx, vy, vsw) = keras.utils.unpack_x_y_sample_weight(
+                    validation_data
+                )
+                validation_data = convert_inputs_to_dataset(
+                    vx, vy, vsw, batch_size
+                )
+
+        return super().fit(
+            x=x,
+            y=None,
+            batch_size=None,
+            sample_weight=None,
+            validation_data=validation_data,
+            **kwargs,
+        )
+
+    def evaluate(
+        self,
+        x=None,
+        y=None,
+        batch_size=None,
+        sample_weight=None,
+        **kwargs,
+    ):
+        # TODO: we can't support a cached eval dataset until we make changes to
```

Comment:
Should we issue a warning?

Reply:
Updated the comment with a little more color.

I don't think we could. This is not something that is exposed to users. We would just be warning every `model.fit()`.

Core Keras does a weird thing during `fit()` where the `x` and `y` arguments to evaluate are ignored, and a secret property containing the validation data is read off. We could maybe rely on this mechanism as is if we apply the preprocessing to our validation data during `fit()`, but then if this caching changed in the future we would break. Felt safest to just disable it.